packages/flutter/lib/src/painting/_network_image_io.dart

```diff
@@ -81,39 +73,115 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
       assert(key == this);
 
       final Uri resolved = Uri.base.resolve(key.url);
-      final HttpClientRequest request = await _httpClient.getUrl(resolved);
-      headers?.forEach((String name, String value) {
-        request.headers.add(name, value);
+
+      final Completer<TransferableTypedData> bytesCompleter = Completer<TransferableTypedData>();
+      final RawReceivePort downloadResponseHandler = RawReceivePort((_DownloadResponse response) {
+        if (response.bytes != null) {
+          if (bytesCompleter.isCompleted) {
+            // If an uncaught error occurred in the worker isolate, we'll have
+            // already completed our bytes completer.
+            return;
+          }
+          bytesCompleter.complete(response.bytes);
+        } else if (response.chunkEvent != null) {
+          chunkEvents.add(response.chunkEvent);
+        } else if (response.error != null) {
+          bytesCompleter.completeError(response.error);
+        } else {
+          assert(false);
+        }
       });
-      final HttpClientResponse response = await request.close();
-      if (response.statusCode != HttpStatus.ok)
-        throw Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
-
-      final Uint8List bytes = await consolidateHttpClientResponseBytes(
-        response,
-        onBytesReceived: (int cumulative, int total) {
-          chunkEvents.add(ImageChunkEvent(
-            cumulativeBytesLoaded: cumulative,
-            expectedTotalBytes: total,
-          ));
-        },
-      );
-      if (bytes.lengthInBytes == 0)
+
+      // This will keep reference to [debugNetworkImageHttpClientProvider] tree-shaken
+      // out of release builds.
+      HttpClientProvider httpClientProvider;
+      assert(() { httpClientProvider = debugNetworkImageHttpClientProvider; return true; }());
+
+      final _DownloadRequest downloadRequest = _DownloadRequest(downloadResponseHandler.sendPort,
+          resolved, headers, httpClientProvider);
+      if (_requestPort != null) {
+        // If worker isolate is properly set up ([_requestPort] is holding
+        // initialized [SendPort]), then just send download request down to it.
+        _requestPort.send(downloadRequest);
+      } else {
+        if (_pendingLoader == null) {
+          // If worker isolate creation was not started, start creation now.
+          assert(_pendingLoadRequests == null);
+          _pendingLoadRequests = <_PendingLoadRequest>[];
+          _pendingLoader = _setupIsolate()..then((Isolate isolate) {
+              final RawReceivePort handleError = RawReceivePort((List<dynamic> errorAndStackTrace) {
```

Comment:
We need to ensure to call `handleError.close()` when the isolate terminates so we don't leak the port.

Reply:
Done